src/irmin/tree.ml

```diff
@@ -1305,18 +1305,12 @@ module Make (P : S.PRIVATE) = struct
     (aux [@tailcall]) [] [ (path, tree) ]
 
   let diff_node (x : node) (y : node) =
-    let bindings n =
-      Node.to_map n >|= function None -> [] | Some m -> StepMap.bindings m
-    in
+    let bindings n = Node.to_map n >|= fun m -> StepMap.bindings (get_ok m) in
     let removed acc (k, (c, m)) =
-      Contents.to_value c >|= function
-      | None -> acc
-      | Some c -> (k, `Removed (c, m)) :: acc
+      Contents.to_value c >|= get_ok >|= fun c -> (k, `Removed (c, m)) :: acc
```

Comment:
it would be nice (but I don't know how much this is feasible) to return a meaningful diff when only one part is missing. Maybe by making the `diff` type a bit more elaborate. Or by skipping dangling hashes, as it used to be the case.